server/data/data-ldap/src/test/java/org/apache/james/user/ldap/ReadOnlyUsersLDAPRepositoryTest.java

```diff
@@ -67,188 +73,235 @@ static void afterAll() {
+                usersRepository.test(JAMES_USER_MAIL, PASSWORD);
+            } catch (Exception e) {
+                LOGGER.info("This exception is expected as we shut down the LDAP and forced its use", e);
+            }
+            DockerLdapSingleton.ldapContainer.unpause();
 
-            assertThat(usersLDAPRepository.supportVirtualHosting()).isTrue();
+            assertThat(usersRepository.test(JAMES_USER_MAIL, PASSWORD)).isTrue();
         }
 
         @Test
-        void supportVirtualHostingShouldReturnFalseWhenReportedInConfig() throws Exception {
-            HierarchicalConfiguration<ImmutableNode> configuration = ldapRepositoryConfiguration(ldapContainer);
-            configuration.addProperty(ReadOnlyUsersLDAPRepository.SUPPORTS_VIRTUAL_HOSTING, "false");
+        void knownUserShouldNotBeAbleToLogInWhenPasswordIsNotCorrectWithVirtualHosting() throws Exception {
+            assertThat(usersRepository.test(JAMES_USER, BAD_PASSWORD)).isFalse();
+        }
 
-            ReadOnlyUsersLDAPRepository usersLDAPRepository = new ReadOnlyUsersLDAPRepository(domainList);
-            usersLDAPRepository.configure(configuration);
+        @Test
+        void unknownUserShouldNotBeAbleToLogInWhenPasswordIsCorrectWithVirtualHosting() throws Exception {
+            assertThat(usersRepository.test(UNKNOWN, PASSWORD)).isFalse();
+        }
 
-            assertThat(usersLDAPRepository.supportVirtualHosting()).isFalse();
+        @Test
+        void specialCharacterInUserInputShouldBeSanitized() throws Exception {
+            Username patternMatchingMultipleUsers = Username.of("j*");
+
+            assertThat(usersRepository.test(patternMatchingMultipleUsers, PASSWORD)).isFalse();
         }
 
         @Test
-        void configureShouldThrowOnNonBooleanValueForSupportsVirtualHosting() {
-            HierarchicalConfiguration<ImmutableNode> configuration = ldapRepositoryConfiguration(ldapContainer);
-            configuration.addProperty(ReadOnlyUsersLDAPRepository.SUPPORTS_VIRTUAL_HOSTING, "bad");
+        void containsWithGetUserShouldBeTrueWithVirtualHosting() throws Exception {
+            assertThat(usersRepository.contains(usersRepository.getUser(JAMES_USER_MAIL.asMailAddress()))).isTrue();
+        }
 
-            ReadOnlyUsersLDAPRepository usersLDAPRepository = new ReadOnlyUsersLDAPRepository(domainList);
+        @Disabled("JAMES-3088 ReadOnlyUsersLDAPRepository.isAdministrator() is case sensitive")
+        @Override
+        @Test
+        public void isAdministratorShouldBeCaseInsentive(TestSystem testSystem) throws Exception {
+        }
 
-            assertThatThrownBy(() -> usersLDAPRepository.configure(configuration))
-                .isInstanceOf(ConversionException.class);
+        @Disabled("JAMES-3088 Users are provisioned by default from Dockerfile, cannot setup this test case")
+        @Override
+        @Test
+        public void listShouldReturnEmptyIteratorWhenEmptyRepository(TestSystem testSystem) {
```

Comment:
you can set it up by having a filter returning 0 user.

Reply:
sorry I don't get it, can you be more specific?

Reply:
there is a `filter` parameter in the ldap users repository. If you set it for example to "uid=unknown", it will only look for users which uid is toto, so it will be like the users repository is empty.

Reply:
Sorry, `list()` doesn't use filter, the only thing available is the restriction config. But IDK how to configure it honestly after spending 30 minutes :(

Reply:
We can set up a separate test class for this

Reply:
you can use an unkwown object class instead

Reply:
(I will look at it)

Reply:
> you can use an unkwown object class instead

Not really if I use an organisation that is not in the ldif file it throws an exception.

Reply:
not an organization, an object class. So that your search filter ends doing objectClass=organization (for example)

Reply:
https://github.com/linagora/james-project/pull/3174/commits/ebce9d8e5ec0e56901a7085e77ce030e3e6d2075

is not satisfying enough?

Reply:
I don't see the need of having a seprated test file for this.
Anyway I have a long term goal of rewritting all this (many tests are missing regarding the ldap repository parameters, so it's not a big deal if merged like this

Reply:
Let's go with your solution for now. I just added a fixup adding some missing tests.